### DISCUSSION TITLE
How to allow to reload custom Keycloak providers and theme during development with Keycloak X deployed as container?

### DISCUSSION THREAD

**Discussion by ghilainm** [keycloak/keycloak#12595] [Archive]
How to allow to reload custom Keycloak providers and themes during development with Keycloak X deployed as a container?

Should we use the two phases (build and start) or only start with auto-build?

Any guidance would be appreciated :).

**Comment by sonOfRa** [comment]
If you're using gradle as a build system, as well as docker-compose to run your local keycloak, you can reload like this:

```groovy
task deploy(type: Copy) {
    into "$projectDir/docker/providers" // adjust this to fit the bind mount to the keycloak providers dir
    from jar
}

task reloadProviders(type: Exec) {
    dependsOn deploy
    workingDir "$projectDir/docker" // adjust workdir to fit the location of your compose-file
    commandLine "docker", "compose", "restart", "keycloak" // can be "docker-compose" instead of "docker" "compose"
}
```

The providers directory is a bind mount into the keycloak providers directory. Together with an entrypoint of `entrypoint: ["/opt/keycloak/bin/kc.sh", "start-dev"]` I was able to get relatively close to the the hot reloading that was possible in the wildfly deployment. Keycloak of course takes a few seconds to restart, but that is manageable.

This has worked well for us so far. Another option might be to hot-reload classes via remote debugging, but that's not something we've tried. Add an environment `DEBUG_PORT='*:8787'`, expose that port to your development host via a port mapping, and reload your code that way.

**Comment by ghilainm** [comment]
Thx a lot for your answer indeed I have the same stack and ended up doing the same kind of thing. I think that would be interesting to make it part of official documentation.
